src/Note.cpp

```diff
@@ -0,0 +1,378 @@
+        (this->*note_fn.value())(ch, std::move(args));
+    } else {
+        ch.send_line("Huh?  Type 'help note' for usage.");
+    }
+}
+
+void NoteHandler::read(Char &ch, ArgParser args) {
+    auto show_note = [&ch](int note_index, const Note &note) {
+        ch.send_line("[{:3}] {}|w: {}|w", note_index, note.sender(), note.subject());
+        ch.send_line("{}|w", note.date());
+        ch.send_line("To: {}|w", note.to_list());
+        if (!note.text().empty())
+            ch.page_to(note.text());
+        ch.send_line("|w");
+        ch.last_note = std::max(ch.last_note, note.date_stamp());
+    };
+
+    if (auto num = args.try_shift_number()) {
+        if (auto *note = notes_.lookup(*num, ch)) {
+            show_note(*num, *note);
+        } else {
+            ch.send_line("No such note.");
+        }
+    } else if (args.empty() || matches_start(args.shift(), "next")) {
+        int note_index;
+        if (auto *note = notes_.lookup(ch.last_note, ch, &note_index)) {
+            show_note(*num, *note);
+        } else {
+            ch.send_line("You have no unread notes.");
+        }
+    } else {
+        ch.send_line("Note read which number?");
+    }
+}
+
+void NoteHandler::list(Char &ch, [[maybe_unused]] ArgParser) {
+    int num = 0;
+    for (auto &note : notes_.notes()) {
+        if (note.is_to(ch)) {
+            bool is_new = note.date_stamp() > ch.last_note && !note.sent_by(ch);
+            ch.send_line("[{:3}{}] {}: {}|w", num, is_new ? "N" : " ", note.sender(), note.subject());
+            num++;
+        }
+    }
+}
+
+void NoteHandler::post(Char &ch, [[maybe_unused]] ArgParser) {
+    if (!ch.pnote) {
+        ch.send_line("You have no note in progress.");
+        return;
+    }
+    if (ch.pnote->to_list().empty()) {
+        ch.send_line("You need to provide a recipient (name, all, or immortal).");
+        return;
+    }
+    if (ch.pnote->subject().empty()) {
+        ch.send_line("You need to provide a subject.");
+        return;
+    }
+    ch.pnote->date(current_time);
+    auto &note = notes_.add(std::move(*ch.pnote));
+    ch.pnote.reset();
+
+    on_change_func_(*this);
+
+    for (auto &chtarg : descriptors().all_but(ch) | DescriptorFilter::to_person()) {
+        if (!IS_SET(chtarg.comm, COMM_NOANNOUNCE) && !IS_SET(chtarg.comm, COMM_QUIET) && note.is_to(chtarg)) {
+            chtarg.send_line("The Spirit of Hermes announces the arrival of a new note.");
+        }
+    }
+    ch.send_line("Ok.");
+}
+
+void NoteHandler::add_line(Char &ch, ArgParser args) {
+    ensure_note(ch).add_line(args.remaining());
+    ch.send_line("Ok.");
+}
+
+void NoteHandler::remove_line(Char &ch, [[maybe_unused]] ArgParser) {
+    if (!ch.pnote || ch.pnote->text().empty()) {
+        ch.send_line("There is no text to delete.");
+        return;
+    }
+    ch.pnote->remove_line();
+    ch.send_line("Ok.");
+}
+
+void NoteHandler::subject(Char &ch, ArgParser args) {
+    ensure_note(ch).subject(args.remaining());
```

Comment:
Hmm,  if someone doesn't set the subject, or uses `note subject <empty>` do we save/send a note without one?

Reply:
No, the `post` code has an error for that (and there's even a test! shock)